server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java

```diff
@@ -127,52 +131,238 @@ void setUpdateFrequency(TimeValue updateFrequency) {
 
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
-        if (event.localNodeMaster() && refreshAndRescheduleRunnable.get() == null) {
-            logger.trace("elected as master, scheduling cluster info update tasks");
-            executeRefresh(event.state(), "became master");
-
-            final RefreshAndRescheduleRunnable newRunnable = new RefreshAndRescheduleRunnable();
-            refreshAndRescheduleRunnable.set(newRunnable);
-            threadPool.scheduleUnlessShuttingDown(updateFrequency, REFRESH_EXECUTOR, newRunnable);
-        } else if (event.localNodeMaster() == false) {
-            refreshAndRescheduleRunnable.set(null);
-            return;
-        }
+        final Runnable newRefresh;
+        synchronized (mutex) {
+            if (event.localNodeMaster() == false) {
+                refreshScheduler = null;
+                return;
+            }
 
-        if (enabled == false) {
-            return;
+            if (refreshScheduler == null) {
+                logger.trace("elected as master, scheduling cluster info update tasks");
+                refreshScheduler = new RefreshScheduler();
+                nextRefreshListeners.add(refreshScheduler.getListener());
+            }
+            newRefresh = getNewRefresh();
+            assert assertRefreshInvariant();
         }
+        newRefresh.run();
 
         // Refresh if a data node was added
         for (DiscoveryNode addedNode : event.nodesDelta().addedNodes()) {
             if (addedNode.isDataNode()) {
-                executeRefresh(event.state(), "data node added");
+                refreshAsync(new PlainActionFuture<>());
                 break;
             }
         }
+    }
+
+    private class AsyncRefresh {
+
+        private final List<ActionListener<ClusterInfo>> thisRefreshListeners;
+        private final CountDown countDown = new CountDown(2);
+
+        AsyncRefresh(List<ActionListener<ClusterInfo>> thisRefreshListeners) {
+            this.thisRefreshListeners = thisRefreshListeners;
+        }
 
-        // Clean up info for any removed nodes
```

Comment:
I simply dropped this "clean-up", it doesn't seem important and didn't always work anyway since it would race against ongoing refreshes too.